docs/topics/coroutines.rst

```diff
@@ -37,8 +37,101 @@ hence use coroutine syntax (e.g. ``await``, ``async for``, ``async with``):
+
+.. _async-start_requests:
+
+Asynchronous start_requests and spider middlewares
+==================================================
+
+.. versionadded:: 2.3
+
+The :meth:`~scrapy.spiders.Spider.start_requests` spider method can be an
+asynchronous generator::
+
+    async def start_requests():
+        # ...
+        yield scrapy.Request(...)
+        # ...
+
+In this case all spider middlewares used with this spider that have the
+:meth:`~scrapy.spidermiddlewares.SpiderMiddleware.process_start_requests`
+method must support this: if they receive an asynchronous iterable, they must
+return one as well. On the other hand, if they receive a normal iterable, they
+shouldn't break and ideally should return a normal iterable too. There can be
+several possible implementations of this.
+
+First, such universal :meth:`process_start_requests` can be an asynchronous
+generator itself, and so it will always convert a normal iterable to an
+asynchronous one. Because a result of a middleware method is passed to the same
+method of the next middleware, it's only possible to mix middlewares with
+synchronous and asynchronous :meth:`process_start_requests` if all synchronous
+ones are called first.
+
+.. autofunction:: scrapy.utils.asyncgen.as_async_generator
+
+Here is an example of a universal middleware using this approach::
+
+    from scrapy.utils.asyncgen import as_async_generator
+
+    class ProcessStartRequestsAsyncGenMiddleware:
+        async def process_start_requests(self, start_requests, spider):
+            async for req in as_async_generator(start_requests):
+                # ... do something with req
+                yield req
+
+If this method includes asynchronous code, that code will work even with
+synchronous :meth:`~scrapy.spiders.Spider.start_requests`.
+
+Another option is to make separate methods for normal and asynchronous
+iterables and choose one at run time::
+
+    from inspect import isasyncgen
+
+    class ProcessStartRequestsAsyncGenMiddleware:
+        def _normal_process_start_requests(self, start_requests, spider):
+            # ... do something with normal start_requests
+
+        async def _async_process_start_requests(self, start_requests, spider):
+            # ... do something with async start_requests
+
+        def process_start_requests(self, start_requests, spider):
+            if isasyncgen(start_requests):
+                return self._async_process_start_requests(start_requests, spider)
+            else:
+                return self._normal_process_start_requests(start_requests, spider)
+
+
+.. _async-start_requests-queue:
+
+Queue behavior with asynchronous start_requests
+===============================================
+
+.. versionadded:: 2.x
+
+By default the requests produced by
+:meth:`~scrapy.spiders.Spider.start_requests` don't follow the usual request
+priorities: they are only scheduled while the downloader queue is empty. In
```

Comment:
I'm not sure this is correct - request.priority is not relevant here, and the logic depends on scheduler queue being empty, not downloader queue. I think requests are scheduled while the downloader queue is not full (because at this point scheduler queue starts to grow), not while it is empty.

Reply:
Agreed.